easybuild/framework/easyblock.py

```diff
@@ -1469,21 +1469,29 @@ def make_module_extra(self, altroot=None, altversion=None):
         for (key, value) in self.cfg['modextravars'].items():
             lines.append(self.module_generator.set_environment(key, value))
 
-        for (key, value) in self.cfg['modextrapaths'].items():
-            if isinstance(value, str):
-                value = [value]
-            elif not isinstance(value, (tuple, list)):
-                raise EasyBuildError("modextrapaths dict value %s (type: %s) is not a list or tuple",
-                                     value, type(value))
-            lines.append(self.module_generator.prepend_paths(key, value, allow_abs=self.cfg['allow_prepend_abs_path']))
-
-        for (key, value) in self.cfg['modextrapaths_append'].items():
-            if isinstance(value, str):
-                value = [value]
-            elif not isinstance(value, (tuple, list)):
-                raise EasyBuildError("modextrapaths_append dict value %s (type: %s) is not a list or tuple",
-                                     value, type(value))
-            lines.append(self.module_generator.append_paths(key, value, allow_abs=self.cfg['allow_append_abs_path']))
+        for name, prepend in [('modextrapaths', True), ('modextrapaths_append', False)]:
+            allow_abs = self.cfg['allow_prepend_abs_path'] if prepend else self.cfg['allow_append_abs_path']
+
+            for (key, value) in self.cfg[name].items():
+                if not isinstance(value, (tuple, list, dict, str)):
+                    raise EasyBuildError(f'{name} dict value "{value}" (type {type(value)}) is not a '
+                                         'list, dict or str')
+                elif isinstance(value, dict):
+                    if 'paths' not in value or 'delimiter' not in value:
+                        raise EasyBuildError(f'{name} dict value "{value}" must contain "paths" and "delimiter"')
+
+                    paths = value['paths']
+                    delim = value['delimiter']
+                    if not isinstance(paths, (list, str)):
+                        raise EasyBuildError('modextrapaths dict value "{value}" paths must be list or str')
+                    if not isinstance(delim, str):
+                        raise EasyBuildError('modextrapaths dict value "{value}" delimiter must be a str')
+                    lines.append(self.module_generator.update_paths(key, paths, prepend=prepend, delim=delim,
+                                                                    allow_abs=allow_abs))
+                else:
+                    if isinstance(value, str):
+                        value = [value]
+                    lines.append(self.module_generator.update_paths(key, value, prepend=prepend, allow_abs=allow_abs))
 
```

Comment:
This seems overly complex with too many type checks. Many of these checks are already done by `update_paths`, so I think we can simplify to:

```suggestion
        for extrapaths_type, prepend in [('modextrapaths', True), ('modextrapaths_append', False)]:
            allow_abs = self.cfg['allow_prepend_abs_path'] if prepend else self.cfg['allow_append_abs_path']

            for (key, value) in self.cfg[extrapaths_type].items():
                if not isinstance(value, (tuple, list, dict, str)):
                    raise EasyBuildError(
                        f"{extrapaths_type} dict value '{value}' (type {type(value)}) is not a 'list, dict or str'"
                    )

                try:
                    paths = value['paths']
                    delim = value['delimiter']
                except (KeyError, TypeError):
                    paths = value
                    delim = ':'

                lines.append(
                    self.module_generator.update_paths(key, paths, prepend=prepend, delim=delim, allow_abs=allow_abs)
                )

```

These code passes unit tests.

Reply:
well ok, though I must admit I don't love the reliance on exception types to implicitly determine the type.
The update_paths part seems unclear what is supposed to take (I would actually prefer if that code instead just took a list like the API docs says, and deal with sanitizing user inputs as early as possible)